### DISCUSSION TITLE
Does anybody run multiple repetitions of one trial?

### DISCUSSION THREAD

**Discussion by MarcoMeter** [optuna/optuna#4377] [Q&A]
Hi!

In the context of Deep Learning, do you run all trials just once for hyperparameter tuning?
What if one trial was just a lucky outlier?
Does optuna consider the repetition of trials?

**Comment by nzw0301** [comment]
Sorry, I cannot fully understand the question, but I'd like to answer the questions.

> In the context of Deep Learning, 

Optuna optimises black-box objectives, so it doesn't care whether the objective uses deep learning. 

> do you run all trials just once for hyperparameter tuning?

Depending on which sampler is used. If either GridSampler or BruteForceSampler is used, yes. Otherwise, not.

> What if one trial was just a lucky outlier?

Sorry, could you explain a little bit more?

> Does optuna consider the repetition of trials?

Yes, it does.

**Reply by MarcoMeter** [reply]
Thanks for your response.

By the context of Deep Learning I refer to trials that have different outcomes depending on the used seed (e.g. initial model weights are sampled). Hence, a great outcome of a trial can be a lucky outlier if further repetitions using a different seed yield terrible results. Especially Deep Reinforcement Learning training runs are very stochastic (agent samples actions, environment is stochastic, ...). One run of a trial is not enough to gain confidence.

**Reply by nzw0301** [reply]
Thanks. So I suppose my answer already addressed your question in such an objective with randomness.
As a toy example, the following case, where the first trial is a lucky outlier with `x=0`, but not optimal value, optina can find the optimal parameter (`x=10`).

```python
import optuna

def o(t):
    if t.number == 0:
        x = t.suggest_int("x", 0, 0)
        return 100000000
    return t.suggest_int("x", 0, 10)
    
s = optuna.create_study(direction="maximize")

s.optimize(o, n_trials=1000)
```

**Reply by MarcoMeter** [reply]
This is roughly my current logic for running repetitions of the same trial:

```python
repeat = 3
period = 500
max_steps = 10000
def o(t):
    x = t.suggest_int("x", 0, 10)
    # Init pseudo trainers with x
    repetitions = ["Trainer_" + str(id) + "_" + str(x) for id in range(repeat)]
    for cycle in range(0, max_steps, period):
        results = []
        for repetition in repetitions:
            results.append(repetition.step()) # step pseudo trainer
        t.report(cycle + period, sum(results) / len(results))
    return sum(results) / len(results)
```

My actual implementation has quite some overhead as I have to move data from CPU  to GPU when switching to another repetition. How can I leverage optuna's API to repeat trials and how does optuna track the results of repeated trials? Are results overwritten?

My main goal is to exploit pruning. I don't want to run another repetition if one of the trial's repetitions has low performance.

One approach could be to run trials without repetition and then repeat the most promising trials.

**Reply by davidmurray** [reply]
Hi @MarcoMeter, I have the same question. I am curious as to what you ended up doing to solve this issue. Thanks
